source/ConditionSet.cpp

```diff
@@ -325,477 +361,372 @@ set<string> ConditionSet::RelevantConditions() const
+					// The parser for the sub-expression handled everything with higher precedence. Start the loop over
+					// to check what this parser needs to do next.
+					continue;
+				}
 
+				// If the expression currently contains a terminal, then push it down.
+				// Also push down the current expression if it has a higher or equal precedence to the new operator.
+				if((children.size() == 0) || (children.size() > 1 && infixOp != expressionOperator &&
+						Precedence(expressionOperator) >= Precedence(infixOp)))
+					if(!PushDownFull(node))
+						return FailParse();
 
-// Evaluate the SubExpression using the given condition maps.
-int64_t ConditionSet::Expression::SubExpression::Evaluate(const ConditionsStore &conditions) const
-{
-	// Sanity check.
-	if(tokens.empty())
-		return 0;
+				// If this expression contains only a single sub-expression, then we can apply the operator directly.
+				if(children.size() == 1)
+					expressionOperator = infixOp;
 
-	// For SubExpressions with no Operations (i.e. simple conditions), tokens will consist
-	// of only the condition or numeric value to be returned as-is after substitution.
-	auto data = SubstituteValues(tokens, conditions);
+				// If we get the same operator as that we had, then let's just process it and continue the loop.
+				if(infixOp == expressionOperator)
+				{
+					++tokenNr;
+					if(!((children.emplace_back()).ParseMini(node, tokenNr)))
+						return FailParse();
 
-	if(!sequence.empty())
-	{
-		// Each Operation adds to the end of the data vector.
-		data.reserve(operatorCount + data.size());
-		for(const Operation &op : sequence)
-			data.emplace_back(op.fun(data[op.a], data[op.b]));
+					continue;
+				}
+				return FailParse(node, "precedence confusion on infix operator");
+			}
+			default:
+				return FailParse(node, "expected infix operator instead of \"" + node.Token(tokenNr) + "\"");
+		}
 	}
-
-	return data.back();
 }
 
 
 
-// Parse the input vector into the tokens and operators vectors. Parentheses are
-// considered simple operators, and also insert an empty string into tokens.
-void ConditionSet::Expression::SubExpression::ParseSide(const vector<string> &side)
+bool ConditionSet::PromoteFirstChild(const DataNode &node)
 {
-	static const string EMPTY;
-	int parentheses = 0;
-	// Construct the tokens and operators vectors.
-	for(size_t i = 0; i < side.size(); ++i)
-	{
-		if(side[i] == "(" || side[i] == ")")
-		{
-			// Ensure reconstruction by adding a blank token.
-			tokens.emplace_back(EMPTY);
-			operators.emplace_back(side[i]);
-			++parentheses;
-		}
-		else if(IsSimple(side[i]))
-		{
-			// Normal operators do not need a token insertion.
-			operators.emplace_back(side[i]);
-			++operatorCount;
-		}
-		else
-			tokens.emplace_back(side[i]);
-	}
+	// Need to copy first, before moving the vector.
+	ConditionSet cs = children[0];
+	expressionOperator = cs.expressionOperator;
+	literal = cs.literal;
+	conditionName = cs.conditionName;
+	children = cs.children;
```

Comment:
Is this essentially a copy assignment operator to the current object? Could that be done using the operator?

Reply:
Yes, it could probably be done in 2 lines, but it is critical to create a copy of `children[0]` before doing the assignment operator to the parent, because otherwise the 'children' array gets emptied before the copy of `children[0]` completes.

Reply:
I'm just wondering if it's worth moving this logic to the operator, just in case it's used somewhere (maybe in a container?)

Reply:
Right, I could just create an assignment operator and let it have the content of this function.

Reply:
Done, but the code is a bit tricky, because the compiler might optimize out the safeguards that prevent `children[0]` from being deleted to early.